src.ts/transaction/qi-transaction.ts

```diff
@@ -84,14 +69,23 @@ export class QiTransaction extends AbstractTransaction<string> implements QiTran
             throw new Error("Cross-shard & cross-ledger transactions are not supported");
         }
 
-        let hash = keccak256(data)
+        let hash = keccak256(this.serialized)
         hash = '0x' + this.originShard+ (originUtxo ? 'F' : '1') + hash.charAt(5) + this.originShard+ (destUtxo ? 'F' : '1') + hash.slice(9)
```

Comment:
The implementation for getting the permuted QIP10 hash of a transaction should, for all intents and purposes, match that of `go-quai`. This means that we should be manipulating the hash as a uint8[] rather than string and setting the values that correspond to the first 4 elements of the uint8[]. Here is the implementation for how `go-quai` handles computing a Qi tx hash:


```go
hash = crypto.Keccak256Hash(data)
// the origin of this tx is the *destination* of the utxos being spent
origin := tx.TxIn()[0].PreviousOutPoint.TxHash[1]
hash[0] = origin
hash[1] |= 0x80 // 10000000 in binary (set first bit to 1)
hash[2] = origin
hash[3] |= 0x80
```